the-collab-lab/tcl-77-smart-shopping-list
# Restrict Navigation Bar Links Based on User Authentication Status

## Description



This code change improves the user experience to reduce confusion on function usability when signed in and out of the application. It dynamically adapts the navbar to only show when a user is logged in and prevents "crafty" users from manually navigating to protected routes and unestablished routes by adding redirects back to the homepage. 

There was some own confusion of being able to navigate to pages when working on other issues and not understanding why I wasn't seeing what I thought there should be only to realize I wasn't logged in. If I was having that issue a new user could land on the application and attempt to navigate and not realize they are missing out because they aren't logged in. We want to provide a good user experience that directs people to the proper use of the application. 

Stumbling upon this issue did make me realize that authentication while handled by firebase still has a UI component that we can think about other than the direct login button. Having the UI change as we have different user states allows for a subconscious understanding (that is fostered by the rest of the web) of how our application should work from the moment of signup.

## Related Issue


Closes #26

## Acceptance Criteria


- [x] The sign in button is on the homepage when no user is logged in
- [x] The sign out button is in the nav bar when a user is logged in and there is no sign in button
- [x] The nav bar is only visible when a user is logged in
- [x] Prevent manual navigation to routes that a user has to be logged into see

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|  ✓    | 💯 Enhancement     |
|  ✓    | ♻️ Refactor |
|  ✓    | 🐛 Bug Fix |

## Updates

### Before



https://github.com/user-attachments/assets/4901eea2-bbf4-4ac0-bd2c-abfbadb5f2e3



### After



https://github.com/user-attachments/assets/145e2aec-ea31-4bc6-a153-8fd579c7cdcd



## Testing Steps / QA Criteria



- From your terminal, pull down this branch with `git pull origin bb/navbar-auth-status` and check that branch out with git checkout `bb-/navbar-auth-status`
- Then `npm ci` to install the newly added dependencies locally and `npm start` to launch the app.
- Open a second terminal to run the typescript compiler using ` tsc --noEmit --watch` to see any type issue.
- If still signed in from previous usage of the app, make sure to sign out and sign back in. 
- When signed out you should only see the sign in button on the homepage with the "hello from the homepage message".
- In the URL bar add `/list` to the end and see that the page will automatically redirect to the homepage if no user is signed in. The same will happen if `/manage-list` is added.
- After seeing the redirect of non-logged in users to the homepage if manually changing the page click the `sign in` button.
- When logged in the nav bar will appear back at the bottom of the page and can be used as it previously was.
- Hit `sign out` and get redirected back to the home page on successful log out.
